shell/common/shell.cc

```diff
@@ -372,6 +371,71 @@ void Shell::NotifyLowMemoryWarning() const {
   // to purge them.
 }
 
+void Shell::RunEngine(RunConfiguration run_configuration) {
+  RunEngine(std::move(run_configuration), nullptr);
+}
+
+void Shell::RunEngine(RunConfiguration run_configuration,
+                      std::function<void(Engine::RunStatus)> result_callback) {
+  FML_DCHECK(is_setup_);
+  FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
+
+  if (!weak_engine_) {
+    result_callback(Engine::RunStatus::Failure);
+  }
+  fml::TaskRunner::RunNowOrPostTask(
+      task_runners_.GetUITaskRunner(),
+      fml::MakeCopyable([run_configuration = std::move(run_configuration),
+                         weak_engine = weak_engine_,
+                         result_callback =
+                             std::move(result_callback)]() mutable {
+        if (!weak_engine) {
+          if (result_callback) {
+            FML_LOG(ERROR)
+                << "Could not launch engine with configuration - no engine.";
+            result_callback(Engine::RunStatus::Failure);
+          }
+          return;
+        }
+        auto result = weak_engine->Run(std::move(run_configuration));
+        if (result == flutter::Engine::RunStatus::Failure) {
+          FML_LOG(ERROR) << "Could not launch engine with configuration.";
+        }
+        if (result_callback) {
+          result_callback(result);
+        }
+      }));
+}
+
+std::optional<int> Shell::GetUIIsolateLastError() const {
+  FML_DCHECK(is_setup_);
+  FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
+
+  if (!weak_engine_) {
+    return std::nullopt;
+  }
+  switch (weak_engine_->GetUIIsolateLastError()) {
```

Comment:
Using a weak pointer on the non-originating thread is not thread safe. See the comment in `weak_ptr.h` that describes safe use. I have pasted the relevant bits below.

Now in this particular case, the engine provides a guarantee that shell subcomponents will be destroyed before its destructor finishes execution on the platform task runner. So, it happens to work out. But I would still be more comfortable if code didn't break safe threading rules because of the lifecycle assumptions. It is not immediately clear from reading just this bit of code about why this is safe even though it breaks rules of safe weak pointer use.

You can use the same pattern used right above in the `RunEngine` call.

```
// Weak pointers are not in general thread-safe. They may only be *used* on a
// single thread, namely the same thread as the "originating" |WeakPtrFactory|
// (which can invalidate the weak pointers that it generates).
//
// However, weak pointers may be passed to other threads, reset on other
// threads, or destroyed on other threads. They may also be reassigned on other
// threads (in which case they should then only be used on the thread
// corresponding to the new "originating" |WeakPtrFactory|).

```

Reply:
Going with the unique_ptr and a comment, as discussed offline